philosophy/FORMS.md

```diff
@@ -0,0 +1,928 @@
+
+#### Enriching received data (e.g. adding additional languages to localized strings)
+
+Let's now assume our product name should be localized. A name can be set for multiple locales. The data looks like this:
+
+```js
+{
+  id: 'party-parrot',
+  version: 1,
+  name: {
+    en: 'Party Parrot',
+    de: 'Party Papagei',
+    es: 'Loro fiesta',
+  },
+}
+```
+
+Imagine the data of our product is a bit older, and the russian locale has been added to the project in the meantime, while the spanish locale was removed.
+Our goal is to show translations for all locales stored on the product itself (even the outdated spanish one), while also showing input fields all locales currently stored on the product. So ultimately we should be showing `en`, `de`, `es`, `ru` in this case.
+
+We can't simply iterate over the languages defined on the project (`en`, `de`, `ru`) as we'd be missing `es`. And we can't simply iterate over the languages stored on the product (`en`, `de`, `es`) as we'd be missing `ru`.
+
+Assume the project's locales are defined as `projectLocales = ['en', 'de', 'ru']`. We have to merge the locales in `docToFormValues`.
+
+```js
+const docToFormValues = (doc, projectLocales) => ({
+  id: doc.id,
+  version: doc.version,
+  name: createLocalizedString(projectLocales, doc.name),
+});
+```
+
+The `createLocalizedString` function ensures that a `{ [locale]: '' }` entry is set for each locale passed as the first argument, and then merges the already defined translations into that object.
+
+This would produce
+
+```js
+{
+  id: 'party-parrot',
+  version: 1,
+  name: {
+    en: 'Party Parrot',
+    de: 'Party Papagei',
+    es: 'Loro fiesta',
+    ru: '',
+  },
+}
+```
+
+Notice that we're now passing additional information (`projectLocales`) to `docToFormValues`. This is a new pattern which allows us to enrich the recieved document. This approach of passing additional information to `docToFormValues` can be used to enrich form values in many ways.
+
+### Validation
+
+#### Translating error messages
+
+During validation, we create error objects like this:
+
+```js
+{
+  values: { name: '' },
+  touched: { name: true },
+  errors: { name: { missing: true } }
+}
+```
+
+As you can see, we use boolean indicators to mark errors for form fields.
+This allows us to determine how to render an error message from the render function,
+instead of having to determine it within the validation function.
+
+For example, we can do
+
+```js
+<Formik
+  render={formik => (
+    <div>
+      <input
+        type="text"
+        name="name"
+        onChange={handleChange}
+        value={formik.values.name}
+      />
+      {formik.touched &&
+        formik.touched.name &&
+        formik.errors &&
+        formik.errors.name &&
+        formik.errors.name.missing && <div>Required field</div>}
+    </div>
+  )}
+/>
+```
+
+> Sidenote: Using the [optional chaining proposal (`?.`)](https://github.com/tc39/proposal-optional-chaining), we can shorten this code to
```

Comment:
not sure if we really need to include this here, this doc is already long

Reply:
Yeah. It felt a bit weird to say this is a nice API when users have to do so many checks before accessing. That's why I wanted to point it out.